### PR TITLE
Add Deduplicating ingest processor

### PR DESCRIPTION
**Pull request by polyfractal**
Adds a processor that deduplicates an array of elements.

Deduplication is aided if the input is sorted, however that means the processor needs to know if the last processor to run was a `sort` processor.  

I've tweaked the Processor interface to include a `setLastType()` method which is invoked by CompoundProcessor.  This allows processors to know what the previous processor was, to help with potential optimization.

We can remove this optimization if we don't want processors to know state about the pipeline.

This is part 2 of the fingerprinting ingest components (the other being #17999).  Part three would be a "meta-processor" that chains split + lowercase + sort + dedupe + join together.  Same concerns in #17999 about a single unified processor vs multiple individual components apply here too.

### CODE REVIEW

**Comment by martijnvg** [comment]
What if we always require the `sort` processor to be defined before this processor? We can then:
* replace the `setLastType()` optimisation with a validation that we check at pipeline construction time by adding a validation method to `Processor.Factory`:
** `default void validate(Pipeline pipeline) {}` By default this does nothing, unless factories implement it. This gives processor implementation like this one to fail if a de-duplication processor isn't preceded by a `sort` processor. (it will need to iterate over all processors and if it finds de-duplication processor then check if the previous processor is a sort processor)
** Add the following logic to `Pipeline.Factory#create(...)` to make sure that validation is being performed:
```java
Pipeline pipeline = new Pipeline(id, description, compoundProcessor);
for (Processor.Factory factory : processorRegistry) {
   factory.validate(pipeline);
}
return pipeline;
```
* remove the de-duplication for unsorted lists.

**Comment by polyfractal** [comment]
Hmm, a potential concern is if a user wants deduplication but not to have their token stream re-arranged by a sort?  I can't think of any off-hand, but it may be a problem?

Also unfortunate that a user might define a pipeline and only know it's "wrong" once they execute and see they need a `sort` too.  Less user-friendly.

How about using that `validate()` method to create a specialized class for each scenario?  If a `sort` has been previously defined, the factory returns a `DeduplicateFromSortedProcessor`, otherwise it returns a `DeduplicateFromUnsortedProcessor` ?  Then we can get rid of the `setLastTag()` runtime call, and each specialized dedupe processor only has one method of operation.

**Comment by martijnvg** [comment]
> Hmm, a potential concern is if a user wants deduplication but not to have their token stream re-arranged by a sort? 

I assumed that users were always ok with sorting the array :) but yes that may not always be desired.

> Also unfortunate that a user might define a pipeline and only know it's "wrong" once they execute and see they need a sort too. Less user-friendly.

But they would find it out when calling the put pipeline api, because the validation is in the factory instead of the processor. I think this ok? The pipeline never gets created and UI's can nicely show this.

> How about using that validate() method to create a specialized class for each scenario?

+1 We can do it easier, I think we don't need an extra method in that case. But if we add `List<Processor` to `Factory#create` then we check what the previously created processors are and pick the right implementation?

**Comment by talevy** [comment]
@polyfractal, what if a user knows beforehand that their array is sorted (pre-pipeline), would they not be hindered by this since the pipeline itself does not have this knowledge?

**Comment by polyfractal** [comment]
@talevy Yep you're right...it'd use the sub-optimal HashSet based method.  OTOH, the current setup guarantees it'll do the right thing.  E.g. if they say it's sorted but not actually, the de-duping would run without error but start removing wrong tokens.

I'm ambivalent on the issue though, we could certainly expose a power-user flag that allows the optimized de-duping if you promise your data is sorted :)

File: core/src/main/java/org/elasticsearch/ingest/processor/DeduplicateProcessor.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        List<?> list = document.getFieldValue(field, List.class);
+        if (list == null) {
+            throw new IllegalArgumentException("field [" + field + "] is null, cannot deduplicate.");
+        }
+
+        if (list.size() <= 1) {
+            return;
+        }
+
+        if (lastType != null && lastType.equals("sort")) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/ingest/processor/DeduplicateProcessor.java]: I was curious about how new ingest processors are implemented, so looked into this PR. Regarding requiring "sort": what if lastType is "sort" but it was sorting another field? Something that might be caught by pipeline validation perhaps.

Comment (polyfractal) [inline core/src/main/java/org/elasticsearch/ingest/processor/DeduplicateProcessor.java]: Oh! That's a really good point.  I'll address it during the validation refactoring stuff, thanks!

**Comment by dakrone** [comment]
@polyfractal are you still working on this?

**Comment by dakrone** [comment]
@polyfractal ping, are you still working on this?

**Comment by polyfractal** [comment]
Nope, I'm the worst and forgot to close this after chatting with @talevy in Prague about it.  Closing!

**Pull request closed without merging**